src/persist-txn/src/txn_read.rs

```diff
@@ -321,19 +340,85 @@ enum WaitTs<T> {
 }
 
 #[derive(Debug)]
-struct TxnsReadTask<T: Timestamp + Lattice + Codec64, C: TxnsCodec = TxnsCodecDefault> {
+struct TxnsReadTask<T: Timestamp + Lattice + Codec64> {
     rx: mpsc::Receiver<TxnsReadCmd<T>>,
-    cache: TxnsCache<T, C>,
+    cache: TxnsCacheState<T>,
+    pending_waits: Vec<PendingWait<T>>,
+    frontier: T,
+}
+
+/// A pending "wait" notification that we will complete once the frontier
+/// advances far enough.
+#[derive(Debug)]
+struct PendingWait<T: Timestamp + Lattice + Codec64> {
+    ts: WaitTs<T>,
+    tx: Option<oneshot::Sender<()>>,
 }
 
-impl<T, C> TxnsReadTask<T, C>
+impl<T: Timestamp + Lattice + Codec64> PendingWait<T> {
+    /// Returns `true` if this [PendingWait] is completed.
+    ///
+    /// A pending wait is completed when the frontier advances far enough or
+    /// when the receiver side hangs up.
+    fn maybe_complete(&mut self, frontier: &T) -> bool {
+        if self.tx.is_none() {
+            // Already completed!
+            return true;
+        }
+
+        match &self.ts {
+            WaitTs::GreaterEqual(ts) => {
+                if frontier >= ts {
+                    let _ = self.tx.take().expect("known to exist").send(());
+                    return true;
+                }
+            }
+            WaitTs::GreaterThan(ts) => {
+                if frontier > ts {
+                    let _ = self.tx.take().expect("known to exist").send(());
+                    return true;
+                }
+            }
+        };
+
+        if let Some(tx) = self.tx.as_ref() {
+            if tx.is_closed() {
```

Comment:
Oooo using tx.is_closed here instead of the Arc::weak I suggested is clever!

One concern I have though is that after we address the `O(n^2)` comment I left below, then these might accumulate in memory. Not sure I have a solution here yet (this is the hard problem that is the reason I hadn't typed this up yet myself)

Reply:
Okay I think I figured something out on my walk! How about this:
- Each wait req is accompanied by a random uuid which makes its way into PendingWait
- A new `TxnsReadCmd::CancelWait{id: uuid::Uuid}` is added
- Whoever sent the wait request is responsible for sending the `CancelWait` on Drop (this gets packed up inside `TxnsRead`)
- Inside the worker we store them as `BTreeSet<(T, Uuid)>` and `HashMap<Uuid, PendingWait>` which I think gets us efficient insert, random-access cancel, and peel-off-times operations